aip/auth/4110.md

```diff
@@ -0,0 +1,153 @@
+
+## Guidance
+
+### Credential Types
+
+This section outlines the supported credential types of the ADC.
+
+#### GCloud CLI Credential
+
+A credential provided by the [GCloud CLI][0] that identifies a human user that
+needs to authenticate to access Google APIs. The auth libraries **must** support
+this credential type.
+
+#### Service Account Key
+
+A credential that identifies a non-human user that needs to authenticate to
+aceess Google APIs. The auth libraries **must** support this
+credential type.
+
+#### OAuth Client ID
+
+A credential that identifies the client application which allows human users to
+sign-in through [3-legged OAuth flow][1], which grants the permissions to the
+application to access Google APIs on behave of the human user. The auth
+libraries **may** support this credential type.
+
+### Environment Variables
+
+The auth libraries **must** support the following environment variables to allow
+developers to provide authentication configuration for their application:
+
+#### GOOGLE\_APPLICATION\_CREDENTIALS
+
+The specified value will be used as the full path for ADC to locate the
+credentials file. The credentials file **should** be one of the following types:
+
+- Service account key
+- GCloud CLI credentials
+
+The credentials __may__ be the OAuth Client ID if it is supported by the auth
+library.
+
+This value __must__ be overridden by any user specified credentials at the
+programming level (e.g. through clien-level options)
+
+#### GOOGLE\_API\_KEY
+
+[API keys][2] are a simple encrypted string that identifies an application,
+which are useful for accessing public data anonymously. The specified value of
+this environment variable **should** be used as the API key for the API
+requests.
+
+This value **must** be overridden by any user specified credentials at the
+programming level (e.g. through client options)
+
+**Note:** This value **must** be ignored if GOOGLE\_APPLICATION\_CREDENTIALS is
+specified.
+
+#### GOOGLE\_API\_USE\_CLIENT\_CERTIFICATE
+
+The specified value **must** be either true or false. The client certificate
+**must** be ignored if this variable is set to false.
+
+### Inputs & Outputs
+
+From the input/output perspective, the inputs of _ADC_ __should__ be the
+credentials as well as the underlying environment that provides these
+credentials. 
+
+For example, the _GOOGLE_APPLICATION_CREDENTIALS_ environment variable can provide the
+default credential JSON as the input here, or the well-known path that gCloud
+uses to store the default user credential JSON. The output is the access token
+that application can use to access the Google APIs. This access token __may__ be
+a bearer token, a certificate bound token or STS depending on the chosen
+authentication flow.
+
+## Expected Behavior
+
+This section outlines the expected behavior of the ADC. Auth libraries **must**
+implement these concepts in order to be considered complete.
+
+1. **Check environment variables**
```

Comment:
Maybe do this in graphviz? (See AIP-1 for an example.)

Reply:
Thanks for the suggestion. It's very helpful.

Reply:
I sort of hoped a graphviz could _replace_ this.

Reply:
Discussed with @broady that we think it is too much information to fit everything in a graph (the logic is quite complicated), it is probably better (and easier to translate to code) in the current text form.